### ISSUE TITLE
CustomFinder that starts a saga throws exception

### ISSUE THREAD

**Issue by danielmarbach** [Particular/NServiceBus#4888]
## Who's affected

* users trying to use a custom finder for a message that starts a saga

## Symptoms

When a custom finder is used for a message that starts a saga, an exception is thrown indicating that a correlated property must have a non default value assigned when a new saga instance is created, even though the saga data property value is assigned in the handle method.

```
System.Exception: The correlated property 'Identifier' on saga 'TestSaga' does not have a value.
 A correlated property must have a non default (i.e. non null and non-empty) value assigned when a new saga instance is created.
```

## Steps to reproduce

* Create the following saga definition and finder

```
public class TestSagaData : ContainSagaData
{
    public long Identifier { get; set; }
}

public class TestSaga : SqlSaga<TestSagaData>,
                        IAmStartedByMessages<FirstCommand>,
                        IHandleMessages<SecondCommand>
{

    protected override string CorrelationPropertyName => nameof(TestSagaData.Identifier);


    protected override void ConfigureMapping(IMessagePropertyMapper mapper)
    {
        // not required because of custom finder
        //mapper.ConfigureMapping<FirstCommand>(m => m.Identifier); 
        mapper.ConfigureMapping<SecondCommand>(m => m.Identifier);
    }

    public async Task Handle(FirstCommand message, IMessageHandlerContext context)
    {
        Data.Identifier = message.Identifier;
    }

    public async Task Handle(SecondCommand message, IMessageHandlerContext context)
    {
        MarkAsComplete();
    }
}

class TestSagaFinder : IFindSagas<TestSagaData>.Using<FirstCommand>
{
    public Task<TestSagaData> FindBy(FirstCommand message, SynchronizedStorageSession storageSession, ReadOnlyContextBag context)
    {
        return Task.FromResult(default(TestSagaData)); // naive implementation just to repro the bug
    }
}
```

* Send a local message

```
await session.SendLocal(new FirstCommand() {Identifier = 42}).ConfigureAwait(false);
```

The following exception is raised

```
System.Exception: The correlated property 'Identifier' on saga 'TestSaga' does not have a value.
A correlated property must have a non default (i.e. non null and non-empty) value assigned when a new saga instance is created.
    at NServiceBus.Sagas.ActiveSagaInstance.ValidateCorrelationPropertyHaveValue() in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Sagas\ActiveSagaInstance.cs:line 174
    at NServiceBus.Sagas.ActiveSagaInstance.ValidateChanges() in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Sagas\ActiveSagaInstance.cs:line 157
    at NServiceBus.SagaPersistenceBehavior.&lt;Invoke&gt;d__1.MoveNext() in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Sagas\SagaPersistenceBehavior.cs:line 156
--- End of stack trace from previous location where exception was thrown ---
    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
    at NServiceBus.LoadHandlersConnector.&lt;Invoke&gt;d__1.MoveNext() in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Pipeline\Incoming\LoadHandlersConnector.cs:line 40
--- End of stack trace from previous location where exception was thrown ---
    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
    at NServiceBus.InvokeSagaNotFoundBehavior.&lt;Invoke&gt;d__0.MoveNext() in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Sagas\InvokeSagaNotFoundBehavior.cs:line 16
--- End of stack trace from previous location where exception was thrown ---
    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
    at NServiceBus.DeserializeLogicalMessagesConnector.&lt;Invoke&gt;d__1.MoveNext() in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Pipeline\Incoming\DeserializeLogicalMessagesConnector.cs:line 33
--- End of stack trace from previous location where exception was thrown ---
    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
    at NServiceBus.InvokeAuditPipelineBehavior.&lt;Invoke&gt;d__1.MoveNext() in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Audit\InvokeAuditPipelineBehavior.cs:line 18
--- End of stack trace from previous location where exception was thrown ---
    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
    at NServiceBus.SubscriptionReceiverBehavior.&lt;Invoke&gt;d__1.MoveNext() in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Routing\MessageDrivenSubscriptions\SubscriptionReceiverBehavior.cs:line 29
--- End of stack trace from previous location where exception was thrown ---
    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
    at NServiceBus.ReceivePerformanceDiagnosticsBehavior.&lt;Invoke&gt;d__2.MoveNext() in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Performance\Statistics\ReceivePerformanceDiagnosticsBehavior.cs:line 40
--- End of stack trace from previous location where exception was thrown ---
    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
    at NServiceBus.ProcessingStatisticsBehavior.&lt;Invoke&gt;d__0.MoveNext() in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Performance\Statistics\ProcessingStatisticsBehavior.cs:line 27
--- End of stack trace from previous location where exception was thrown ---
    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
    at NServiceBus.TransportReceiveToPhysicalMessageProcessingConnector.&lt;Invoke&gt;d__1.MoveNext() in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Pipeline\Incoming\TransportReceiveToPhysicalMessageProcessingConnector.cs:line 37
--- End of stack trace from previous location where exception was thrown ---
    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
    at NServiceBus.MainPipelineExecutor.&lt;Invoke&gt;d__1.MoveNext() in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Pipeline\MainPipelineExecutor.cs:line 32
--- End of stack trace from previous location where exception was thrown ---
    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
    at NServiceBus.ReceiveStrategy.&lt;TryProcessMessage&gt;d__7.MoveNext() in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Transports\Msmq\ReceiveStrategy.cs:line 109
--- End of stack trace from previous location where exception was thrown ---
    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
    at NServiceBus.TransactionScopeStrategy.&lt;ProcessMessage&gt;d__2.MoveNext() in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Transports\Msmq\TransactionScopeStrategy.cs:line 86
```

I tested this with SQL Persistence but I'm pretty sure the issue is in the core. For more details see the Analysis section.

This prevents existing customers to upgrade from versions prior to v6 since in v4 and v5 this seems to work (have not double checked but tried to reason about the code).

Related case https://eu1.salesforce.com/500D000001Vc9XF

### Normal flow with mapped properties
- Message arrives
- New saga entity gets created with `CreateNewSagaEntity` and sets the property of the saga entity
- `AttachNewEntity` on the active saga instance reads the value back from the entity and uses it to create its internal correlation property info object as a field

Outcome: The correlation property of the active saga instance already contains the correct value eventhough the instance was newly created. Assigned and instantion of the required data happens before the `Handle` method is executed. Therefore when `ValidateChanges` is called on `ActiveSagaInstance` it will work because the ActiveSagaInstance correlation property will have a non default value

### Flow with custom finder

- Message arrives
- New saga entity gets created with `CreateNewSagaEntity` but there is nothing to be set on the saga entity because we have now mapping definition
- `AttachNewEntity` will find a correlation property definition on the metadata (because of the second mapping) but the value on the entity will be the default value therefore the correlation property field inside `ActiveSagaInstance` will represent the default value of the property

Outcome: When `ValidateChanges` is called on `ActiveSagaInstance` it will throw because the field will still have a correlation property with the default value although the user code might have set the correlation property to a non-default value.

The code path seems to assume that the message that starts a saga always has a mapping and is not using a custom finder. 

### Flow with custom finders for all properties (no mapping)

Currently not possible because of the behavior described in Flow with custom finder.

## Hack to make it work

Call

```
sagaInstanceState.AttachExistingEntity(sagaInstanceState.Instance.Entity);
```

before `ValidateChanges`

**Comment by danielmarbach** [comment]
Of course the above described hack might have side effects as well. Another option would be no no longer validate whether the correlation property has a non default value. One could argue the check is not really useful anyway because:

When the property finder is executed the correlation property will always have the right value. When uses would reset the correlation on the saga instance the active saga instance correlation property would still have the previous value. Maybe a proper check on the actual saga entity would be better suited then checking the internal correlation property value in the active saga instance.

**Comment by marktaling** [comment]
I can confirm this behavior works fine with NServiceBus 4, as it is our current production scenario. Of course, this particular issue cannot arise in NServiceBus 4, as the correlation property did not yet exist in that version.

**Comment by dbelcham** [comment]
Fixed in 6.3.6

**Comment by danielmarbach** [comment]
Thanks. Notified Mark and closed the case
